### PR TITLE
[tune] Add SigOpt Integration

### PR DESCRIPTION
**Pull request by andrewztan**
## What do these changes do?

Integrates Sigopt into Tune as a search algorithm.

### CODE REVIEW

**Comment by richardliaw** [comment]
Previous PR here: https://github.com/ray-project/ray/pull/1570/files

**Comment by richardliaw** [comment]
@andrewztan, are you planning to do this? can you push what you have?

**Comment by andrewztan** [comment]
Added all functions except on_trial_result

File: python/ray/tune/suggest/sigopt.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    Requires SigOpt to be installed. Can only support 10 concurrent trials
+    at once, or whatever is allowed by the user's SigOpt plan.
+
+    Parameters:
+        space (list of dict): SigOpt configuration. Parameters will be sampled
+            from this configuration and will be used to override
+            parameters generated in the variant generation process.
+        max_concurrent (int): Number of maximum concurrent trials. Defaults
+            to 10.
```
Comment (richardliaw) [inline python/ray/tune/suggest/sigopt.py]: 1?

File: python/ray/tune/examples/sigopt_example.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+"""This test checks that SigOpt is functional.
```
Comment (richardliaw) [inline python/ray/tune/examples/sigopt_example.py]: can you make sure this file is run in the multinode tests? see the other PR for example

**Review by richardliaw — COMMENTED** [review]
I think we just need to add this to tests and add documentation; see the bayesian opt PR and lets try to get this merged by tomorrow!

File: python/ray/tune/examples/sigopt_example.py (hunk 2 above)
Comment (andrewztan) [reply python/ray/tune/examples/sigopt_example.py]: 👍

File: python/ray/tune/suggest/sigopt.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                          error=False,
+                          early_terminated=False):
+        """Passes the result to SigOpt unless early terminated or errored.
+
+        Creates SigOpt Observation object for trial.
+        """
+        if result:
+            self.conn.experiments(self.experiment.id).observations().create(
+                suggestion=self._live_trial_mapping[trial_id].id,
+                value=result[self._reward_attr],
```
Comment (richardliaw) [inline python/ray/tune/suggest/sigopt.py]: does SigOpt always maximize?

File: python/ray/tune/suggest/sigopt.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            )
+            # Update the experiment object
+            self.experiment = self.conn.experiments(self.experiment.id).fetch()
```
Comment (richardliaw) [inline python/ray/tune/suggest/sigopt.py]: what happens if a trial errored or failed, do you have to do any special handling on the sigopt side?

File: python/ray/tune/suggest/sigopt.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    sgo = None
+
+from ray.tune.suggest.suggestion import SuggestionAlgorithm
+
+
+class SigOptSearch(SuggestionAlgorithm):
+    """A wrapper around SigOpt to provide trial suggestions.
+
+    Requires SigOpt to be installed. 
+    The number of concurrent trials supported depends on the user's
```
Comment (richardliaw) [inline python/ray/tune/suggest/sigopt.py]: Can move this line into Parameters

**Review by richardliaw — COMMENTED** [review]
Last comments; then we can probably merge after

File: python/ray/tune/suggest/sigopt.py (hunk 3 above)
Comment (andrewztan) [reply python/ray/tune/suggest/sigopt.py]: SigOpt maximizes real-valued objective functions by default

File: python/ray/tune/suggest/sigopt.py (hunk 5 above)
Comment (andrewztan) [reply python/ray/tune/suggest/sigopt.py]: 👍

File: python/ray/tune/suggest/sigopt.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            max_concurrent=1,
+            reward_attr="episode_reward_mean",
+            **kwargs):
+        assert sgo is not None, "SigOpt must be installed!"
+        assert type(max_concurrent) is int and max_concurrent > 0
+        self._max_concurrent = max_concurrent
+        self._reward_attr = reward_attr
+        self._live_trial_mapping = {}
+
+        self.conn = sgo.Connection(client_token=os.environ['SIGOPT_KEY'])
```
Comment (richardliaw) [inline python/ray/tune/suggest/sigopt.py]: can you actually document this?

File: python/ray/tune/examples/sigopt_example.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        reporter(
+            timesteps_total=i,
+            neg_mean_loss=-(config["height"] - 14)**2 +
+            abs(config["width"] - 3))
+        time.sleep(0.02)
+
+
+if __name__ == '__main__':
+    import argparse
+
```
Comment (richardliaw) [inline python/ray/tune/examples/sigopt_example.py]: in this example, can you also add a check for the key? (like an assertion)

File: doc/source/tune-searchalg.rst [hunk 8]
```diff
@@ -73,6 +74,29 @@ An example of this can be found in `hyperopt_example.py <https://github.com/ray-
     :show-inheritance:
     :noindex:
 
+SigOpt Search
+-------------
+
+The ``SigOptSearch`` is a SearchAlgorithm that is backed by `SigOpt <https://sigopt.com/>`__ to perform sequential model-based hyperparameter optimization. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune's default variant generation/search space declaration when using SigOptSearch.
+
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]: Can you make a note about the SIGOPT_KEY here?

**Review by richardliaw — COMMENTED** [review]
Last comments about API key

File: python/ray/tune/examples/sigopt_example.py (hunk 7 above)
Comment (andrewztan) [reply python/ray/tune/examples/sigopt_example.py]: should i make a check for the key in sigopt.py too?

File: test/jenkins_tests/run_multi_node_tests.sh [hunk 9]
```diff
@@ -350,12 +350,16 @@ docker run --rm --shm-size=${SHM_SIZE} --memory=${MEMORY_SIZE} $DOCKER_SHA \
...
 
+docker run --rm --shm-size=${SHM_SIZE} --memory=${MEMORY_SIZE} $DOCKER_SHA \
+    python /ray/python/ray/tune/examples/bayesopt_example.py \
+    --smoke-test
+
 docker run --rm --shm-size=${SHM_SIZE} --memory=${MEMORY_SIZE} $DOCKER_SHA \
     python /ray/python/ray/tune/examples/hyperopt_example.py \
     --smoke-test
 
 docker run --rm --shm-size=${SHM_SIZE} --memory=${MEMORY_SIZE} $DOCKER_SHA \
```
Comment (richardliaw) [inline test/jenkins_tests/run_multi_node_tests.sh]:
```suggestion
docker run --rm --shm-size=${SHM_SIZE} --memory=${MEMORY_SIZE} -e SIGOPT_KEY $DOCKER_SHA \
```

**Review by richardliaw — APPROVED** [review]

**Pull request merged**